controller/konnect/ops/ops.go

```diff
@@ -99,6 +105,40 @@ func Create[
 		return nil, fmt.Errorf("unsupported entity type %T", ent)
 	}
 
+	var errConflict *sdkkonnecterrs.ConflictError
+	switch {
+	case errors.As(err, &errConflict):
+		switch ent := any(e).(type) {
+		case *konnectv1alpha1.KonnectGatewayControlPlane:
+			id, err = getControlPlaneForUID(ctx, sdk.GetControlPlaneSDK(), ent)
+			// ---------------------------------------------------------------------
+			// TODO: add other Konnect types
+		default:
+
+		}
+
+		if err == nil && id != "" {
+			e.SetKonnectID(id)
+			SetKonnectEntityProgrammedCondition(e)
+		} else {
+			SetKonnectEntityProgrammedConditionFalse(e, consts.KonnectEntitiesFailedToCreateReason, err.Error())
+		}
+	case err != nil:
+		if id != "" {
+			e.SetKonnectID(id)
+		}
```

Comment:
Should we expect the `id` to be non-empty if the error is not `nil`? It looks like breaking the convention of not returning a valid value when an error is returned.

Reply:
So this is somewhat of a special case when we create a resource and then set associations like we do for CPs with group membership: https://github.com/Kong/gateway-operator/blob/9093830f84aedfe5e96aa3c812aef72d8f2c541c/controller/konnect/ops/ops_controlplane.go#L88-L90.

So the object at this point would exist but there would be an error that we'd surface in its conditions. Do you think we can improve this somehow?

Reply:
We discussed this on a call, and the conclusion was to change the `[create|update]ControlPlane` signatures to not return an ID and a condition reason, but instead return a specific error type when the relations' setting fails and handle it in the top-level `Create` or Update` functions accordingly (set the ID on an object and propagate the `Programmed` condition with `False` and the returned reason).